### PR TITLE
VULCAN-85 The Representor

### PR DESCRIPTION
**Pull request by ahdezma**

### CODE REVIEW

**Comment by ahdezma** [comment]
ci:retest

File: vulcan-api/src/main/java/com/liferay/vulcan/identifier/LongIdentifier.java [hunk 1]
```diff
@@ -33,8 +33,8 @@ public default String getId() {
 	 *
 	 * <p>
 	 * For example with a resource obtained in the URL:
-	 * <code>/p/customer/2012/product/100</code> this method will return
-	 * <code>100</code> as a <code>long</code> value.
+	 * <code>/p/product/100</code> this method will return <code>100</code> as a
```
Comment (nhpatt) [inline vulcan-api/src/main/java/com/liferay/vulcan/identifier/LongIdentifier.java]: Still no @review?

File: vulcan-api/src/main/java/com/liferay/vulcan/resource/RelatedModel.java [hunk 2]
```diff
@@ -12,7 +12,7 @@
  * details.
  */
 
-package com.liferay.vulcan.wiring.osgi.model;
+package com.liferay.vulcan.resource;
```
Comment (nhpatt) [inline vulcan-api/src/main/java/com/liferay/vulcan/resource/RelatedModel.java]: 👏

File: vulcan-application/src/main/java/com/liferay/vulcan/application/internal/endpoint/RootEndpointImpl.java [hunk 3]
```diff
@@ -104,17 +104,24 @@
...
 
-		Try<Routes<T>> routesTry = _getRoutesTry(path);
+		Class<T> modelClass = _resourceManager.getModelClass(path);
 
-		return routesTry.map(
-			Routes::getBinaryFunctionOptional
-		).map(
-			Optional::get
-		).mapFailMatching(
+		Optional<Representor<T, Identifier>> representorOptional =
```
Comment (nhpatt) [inline vulcan-application/src/main/java/com/liferay/vulcan/application/internal/endpoint/RootEndpointImpl.java]: Shouldn't be the method representorOptional?

File: vulcan-jaxrs-json/src/main/java/com/liferay/vulcan/jaxrs/json/internal/writer/SingleModelMessageBodyWriter.java [hunk 4]
```diff
@@ -179,47 +181,56 @@ public void writeTo(
...
+					stream -> stream.forEach(
+						relatedCollection -> _writeRelatedCollection(
+							singleModelMessageMapper, jsonObjectBuilder,
+							relatedCollection, singleModel,
+							embeddedPathElements, fields)));
+
+				Optional<Representor<V, Identifier>> representorOptional =
+					_resourceManager.getRepresentor(modelClass);
+
+				representorOptional.ifPresent(
```
Comment (nhpatt) [inline vulcan-jaxrs-json/src/main/java/com/liferay/vulcan/jaxrs/json/internal/writer/SingleModelMessageBodyWriter.java]: liferay formatting makes this review difficult :(

File: vulcan-jaxrs-json/src/main/java/com/liferay/vulcan/jaxrs/json/internal/writer/SingleModelMessageBodyWriter.java [hunk 5]
```diff
@@ -270,48 +281,59 @@ public void writeTo(
...
-				jsonObjectBuilder, field, value));
-
 		_writerHelper.writeTypes(
 			modelClass,
 			types -> singleModelMessageMapper.mapTypes(
 				jsonObjectBuilder, types));
 
-		Optional<String> optional = _writerHelper.getSingleURLOptional(
-			singleModel, _httpServletRequest);
+		Optional<Stream<RelatedCollection<U, ?>>> relatedCollectionsOptional =
```
Comment (nhpatt) [inline vulcan-jaxrs-json/src/main/java/com/liferay/vulcan/jaxrs/json/internal/writer/SingleModelMessageBodyWriter.java]: Stream?

File: vulcan-api/src/main/java/com/liferay/vulcan/identifier/LongIdentifier.java (hunk 1 above)
Comment (ahdezma) [reply vulcan-api/src/main/java/com/liferay/vulcan/identifier/LongIdentifier.java]: I've added all of them in the last pull-request :)

File: vulcan-api/src/main/java/com/liferay/vulcan/resource/RelatedModel.java (hunk 2 above)
Comment (ahdezma) [reply vulcan-api/src/main/java/com/liferay/vulcan/resource/RelatedModel.java]: :)

File: vulcan-application/src/main/java/com/liferay/vulcan/application/internal/endpoint/RootEndpointImpl.java (hunk 3 above)
Comment (ahdezma) [reply vulcan-application/src/main/java/com/liferay/vulcan/application/internal/endpoint/RootEndpointImpl.java]: https://github.com/liferay/com-liferay-vulcan/pull/67/commits/d4df90e64b243db2e942d1e5a90009ddd887e367

File: vulcan-jaxrs-json/src/main/java/com/liferay/vulcan/jaxrs/json/internal/writer/SingleModelMessageBodyWriter.java (hunk 5 above)
Comment (ahdezma) [reply vulcan-jaxrs-json/src/main/java/com/liferay/vulcan/jaxrs/json/internal/writer/SingleModelMessageBodyWriter.java]: You mean calling it `streamOptional`?

File: vulcan-jaxrs-json/src/main/java/com/liferay/vulcan/jaxrs/json/internal/writer/SingleModelMessageBodyWriter.java (hunk 4 above)
Comment (ahdezma) [reply vulcan-jaxrs-json/src/main/java/com/liferay/vulcan/jaxrs/json/internal/writer/SingleModelMessageBodyWriter.java]: :(

**Comment by ahdezma** [comment]
ci:retest

File: vulcan-jaxrs-json/src/main/java/com/liferay/vulcan/jaxrs/json/internal/writer/SingleModelMessageBodyWriter.java (hunk 5 above)
Comment (nhpatt) [reply vulcan-jaxrs-json/src/main/java/com/liferay/vulcan/jaxrs/json/internal/writer/SingleModelMessageBodyWriter.java]: I meant that it is strange to see a stream as a type

**Pull request merged**